vowpalwabbit/cb_explore_adf_common.cc

```diff
@@ -0,0 +1,102 @@
+#include "cb_explore_adf_common.h"
+#include "cb_adf.h"
+
+namespace VW
+{
+namespace cb_explore_adf
+{
+cb_explore_adf_base::~cb_explore_adf_base()
+{
+  m_gen_cs.pred_scores.costs.delete_v();
+}
+
+size_t cb_explore_adf_base::fill_tied(v_array<ACTION_SCORE::action_score>& preds)
+{
+  if (preds.size() == 0)
+    return 0;
+  size_t ret = 1;
+  for (size_t i = 1; i < preds.size(); ++i)
+    if (preds[i].score == preds[0].score)
+      ++ret;
+    else
+      return ret;
+  return ret;
+}
+
+void cb_explore_adf_base::output_example(vw& all, multi_ex& ec_seq)
```

Comment:
I'm not sure output belongs in the class. It maintains the all dependency and is pretty much a command line driver requirement

Reply:
output uses one of the data members held in the class, its also common for all of the reductions here. An alternative would be to use a mixin class to add the functionality while being separate

Reply:
Is a free function able to interrogate all of the information needed from the class to be able to do this?

Reply:
no, output_example requires the m_gen_cs member. For now its public (not even protected for now), but ideally it would be hidden as a private member of the base